doc/keymap-format-text-v1.md

```diff
@@ -1166,13 +1600,107 @@ In this example, the keycode `<AD01>` produces two symbols on level 1
+index the [symbols table] presented in the previous section.
+
+A key type is set using the following syntax:
+
+<!-- TODO: it also works without the group index -->
+
+```c
+key <AD01> {
+    type[Group1] = "TWO_LEVEL", // Type
+    [q, Q]                      // Symbols
+};
+```
 
-@todo how to bind key actions
+The name of the key type is enclosed between double quotes.
+
+The key type may be omitted and will default to:
+
+- `key.type` global defaults, if set.
+- a standard type using the following **heuristic**:
+  <!-- See: xkbcomp/symbols.c (FindAutomaticType) -->
+  - **1** keysym: `ONE_LEVEL`
+  - **2** keysyms:
+    - if the two keysyms are letter and the first is lower case and the other
+      upper case, then [`ALPHABETIC`][ALPHABETIC];
+    - if one of the keysyms is numpad, then `KEYPAD` else [`TWO_LEVEL`][TWO_LEVEL].
+  - **3 or 4** keysyms (a missing 4th keysym is set to `NoSymbol`):
+    - if the first two keysyms are letters and the first is lower case and the
+      other upper case:
+      - if the last two keysyms are letters and the first is lower case and the
+        other upper case then [`FOUR_LEVEL_ALPHABETIC`][FOUR_LEVEL_ALPHABETIC];
+      - else [`FOUR_LEVEL_SEMIALPHABETIC`][FOUR_LEVEL_SEMIALPHABETIC].
+    - if one of the first two keysyms is numpad, then `FOUR_LEVEL_KEYPAD`;
+    - else [`FOUR_LEVEL`][FOUR_LEVEL].
+
+@figure@figcaption
+Commented examples for inferred types:
+@endfigcaption
+
+```c
+// 1 to 2 keysyms
+key <LFSH> { [Shift_L] };                    // Type: ONE_LEVEL
+key <AE01> { [1, exclam] };                  // Type: TWO_LEVEL
+key <AD01> { [q, Q] };                       // Type: ALPHABETIC
+key <KP1>  { [KP_End, KP_1] };               // Type: KEYPAD
+// Edge case: this is consider alphabetic, although
+// the lower case does not correspond to the upper case.
+key <AD01> { [q, N] };                       // Type: ALPHABETIC
+
+// 3 to 4 keysyms
+key <AE01> { [1, exclam, bar] };             // Type: FOUR_LEVEL
+key <AE01> { [1, exclam, bar, exclamdown] }; // Type: FOUR_LEVEL
+key <AD01> { [q, Q, at] };                   // Type: FOUR_LEVEL_SEMIALPHABETIC
+key <AD01> { [q, Q, at, Greek_OMEGA] };      // Type: FOUR_LEVEL_SEMIALPHABETIC
+key <AD05> { [t, T, tslash, Tslash] };       // Type: FOUR_LEVEL_ALPHABETIC
+
+// The inferred type is `FOUR_LEVEL`, but using `LevelThree+Lock`
+// will produce `Q`, because of the keysyms transformations and
+// the corresponding internal capitalization processing.
+key <AE01> { [1, exclam, q, Q] };            // Type: FOUR_LEVEL
+
+// Won’t work, because there is no heuristic for more than 4 keysyms
+// It will trigger the warnings XKB-183 and XKB-516 and default to ONE_LEVEL,
+// ignoring all the keysyms but the first one.
+key <AD01> {[q, Q, at, any, masculine, U2642]};
+// Will work as expected
+key <AD01> {
+    type[Group1] = "EIGHT_LEVEL_SEMIALPHABETIC",
+    [q, Q, at, any, masculine, U2642]
+};
+```
+@endfigure
```

Comment:
this example further clarifies the explanation, nice !

Reply:
Thanks